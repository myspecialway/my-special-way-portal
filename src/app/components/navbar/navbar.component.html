<nav class="navbar navbar-expand-lg navbar-white  navbar-absolute fixed-top">
  <div class="container-fluid">
  {{ 'NAVBAR.HELLO' | translate}}  
    <div class="navbar-wrapper">
      <a class="navbar-brand" href="https://www.tlv-edu.gov.il/sites/yachdav/Pages/Vision.aspx">
        <img class="logo-img" src="/assets/img/msw-logo.png"/>
        <img class="logo-img-tlv" src="/assets/img/tlv-logo-small-white.jpg"/>

      </a>
      <div class="msw-page-title-dropdown dropdown">
        <a class="" href="http://example.com" id="navbarDropdownMenuLink0" data-toggle="dropdown"
           aria-haspopup="true" aria-expanded="false">
            <span class="msw-page-title">
              <span>{{getSelectedMenuItem()}}</span>
              <i class="material-icons">arrow_drop_down</i>
            </span>
          <!--<span class="notification">5</span>-->
        </a>
        <div class="dropdown-menu dropdown-menu-link" aria-labelledby="navbarDropdownMenuLink">
            <div routerLinkActive="active" *ngFor="let menuItem of menuItems" class="{{menuItem.class}} nav-item"  data-test-id="header-nav-links">
              <a class="dropdown-item" [routerLink]="[menuItem.path]" (click)="selectMenuItem(menuItem)">
                {{menuItem.title}}
              </a>
            </div>

        </div>
      </div>
    </div>
    <!--<button mat-raised-button class="navbar-toggler" type="button">-->
    <!--<span class="sr-only">Toggle navigation</span>-->
    <!--<span class="navbar-toggler-icon icon-bar"></span>-->
    <!--<span class="navbar-toggler-icon icon-bar"></span>-->
    <!--<span class="navbar-toggler-icon icon-bar"></span>-->
    <!--</button>-->
    <div class="collapse navbar-collapse justify-content-end" id="navigation">
      <!--<form class="navbar-form">-->
      <!--<div class="input-group no-border">-->
      <!--<input type="text" value="" class="form-control" placeholder="Search...">-->
      <!--<button mat-raised-button type="submit" class="btn btn-white btn-round btn-just-icon">-->
      <!--<i class="material-icons">search</i>-->
      <!--<div class="ripple-container"></div>-->
      <!--</button>-->
      <!--</div>-->
      <!--</form>-->
      <ul *ngIf="currentUser" class="navbar-nav">
        <li class="nav-item dropdown">
          <a class="nav-link" id="navbarDropdownMenuLink" data-toggle="dropdown"
             aria-haspopup="true" aria-expanded="false">
            <span class="msw-header-user"> 
              שלום
            <span class="msw-header-user-name">{{currentUser}}</span>
            <i class="material-icons">arrow_drop_down</i>
            </span>

            <!--<span class="notification">5</span>-->
            <p>
              <span class="d-lg-none d-md-block">Some Actions</span>
            </p>
          </a>
          <div class="dropdown-menu dropdown-menu-link" aria-labelledby="navbarDropdownMenuLink">
            <a class="dropdown-item" id="logout" [routerLink]="['/login']">{{ 'NAVBAR.LOGOUT' | translate}}  </a>
          </div>
        </li>
            <!-- <div *ngIf="currentUser">Hi, {{currentUser}}</div> -->
      </ul>
    </div>
  </div>
</nav>

<!--
<nav class="navbar navbar-transparent navbar-absolute">
    <div class="container-fluid">
        <div class="navbar-header">
            <button mat-raised-button type="button" class="navbar-toggle" data-toggle="collapse" (click)="sidebarToggle()">
                <span class="sr-only">Toggle navigation</span>
                <span class="icon-bar"></span>
                <span class="icon-bar"></span>
                <span class="icon-bar"></span>
            </button>
            <a class="navbar-brand" href="#">{{getTitle()}}</a>
        </div>
        <div class="collapse navbar-collapse">
            <ul class="nav navbar-nav navbar-right">
                <li>
                    <a href="#pablo" class="dropdown-toggle" data-toggle="dropdown">
                        <i class="material-icons">dashboard</i>
                        <p class="hidden-lg hidden-md">Dashboard</p>
                    </a>
                </li>
                <li class="dropdown">
                    <a href="#" class="dropdown-toggle" data-toggle="dropdown">
                        <i class="material-icons">notifications</i>
                        <span class="notification">5</span>
                        <p class="hidden-lg hidden-md">Notifications</p>
                    </a>
                    <ul class="dropdown-menu">
                        <li><a href="#">Mike John responded to your email</a></li>
                        <li><a href="#">You have 5 new tasks</a></li>
                        <li><a href="#">You're now friend with Andrew</a></li>
                        <li><a href="#">Another Notification</a></li>
                        <li><a href="#">Another One</a></li>
                    </ul>
                </li>
                <li>
                    <a href="#pablo" class="dropdown-toggle" data-toggle="dropdown">
                       <i class="material-icons">person</i>
                       <p class="hidden-lg hidden-md">Profile</p>
                    </a>
                </li>
            </ul>
            <form class="navbar-form navbar-right" role="search">
                <div class="form-group form-black is-empty">
                    <input type="text" class="form-control" placeholder="Search">
                    <span class="material-input"></span>
                </div>
                <button mat-raised-button type="submit" class="btn btn-white btn-round btn-just-icon">
                    <i class="material-icons">search</i><div class="ripple-container"></div>
                </button>
            </form>
        </div>
    </div>
</nav> -->
