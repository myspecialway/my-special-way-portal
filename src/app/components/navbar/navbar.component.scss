@import '../../../assets/scss/core/variables';

:host {
  .dropdown-item {
    mat-icon {
      margin: 0 10px;
    }
  }

  .logo-title {
    color: $msw-primary;
    font-weight: 500;
    margin: 0 10px;
  }
}

.navbar {
  .navbar-brand {
    display: flex;
    flex-direction: row;
    .logo-img {
      height: 36px;
      // width: 36px;
    }
  }

  .collapse .navbar-nav .nav-item .nav-link {
    cursor: pointer;
    .msw-header-user {
      font-size: 16px;
    }
    .msw-header-user-name {
      color: $msw-primary;
    }
  }

  .msw-page-title-dropdown {
    padding-top: 10px;
    .msw-page-title {
      border-right: 1px solid $grey-300;
      color: $grey-900;
      font-size: 18px;
      font-weight: 400;
      letter-spacing: 0.3px;
      padding: 0 20px;
    }

    // .dropdown-menu {
    //   border-radius: 0;
    //   width: 240px;
    // }

    // .dropdown-item {
    //   font-size: 16px;
    //   font-weight: 500;
    //   line-height: 1.25;
    // }

    // .dropdown-menu .dropdown-item:hover,
    // .dropdown-menu .dropdown-item:focus,
    // .dropdown-menu a:hover,
    // .dropdown-menu a:focus,
    // .dropdown-menu a:active {
    //   -webkit-box-shadow: 0 4px 20px 0px rgba(0, 0, 0, 0), 0 7px 10px -5px rgba(106, 0, 207, 0);
    //   box-shadow: 0 4px 20px 0px rgba(0, 0, 0, 0), 0 7px 10px -5px rgba(106, 0, 207, 0);
    //   background-color: transparent;
    //   color: $purple-msw;
    // }
  }
}
