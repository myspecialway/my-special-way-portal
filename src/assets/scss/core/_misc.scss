body {
  background-color: #fafafa;
  color: $black-color;
  font-weight: 300;
}

legend {
  border-bottom: 0;
}

.serif-font {
  font-family: $font-family-serif;
}

// Prevent highlight on mobile
* {
  -webkit-tap-highlight-color: rgba(255, 255, 255, 0);
  -webkit-tap-highlight-color: transparent;
  &:focus {
    outline: 0;
  }
}

a {
  color: $link-color;
  &:hover,
  &:focus {
    color: darken($link-color, 5%);
    text-decoration: none;
  }

  &.text-info {
    &:hover,
    &:focus {
      color: darken($brand-info, 5%);
    }
  }

  & .material-icons {
    vertical-align: middle;
  }
}

.form-check,
label {
  font-size: 14px;
  line-height: 1.42857;
  color: $checkboxes-text-color;
  font-weight: 400;
}

/*           Animations              */
.animation-transition-general {
  @include transition-all($general-transition-time, $transition-linear);
}

.animation-transition-slow {
  @include transition-all($slow-transition-time, $transition-linear);
}

.animation-transition-fast {
  @include transition-all($fast-transition-time, $transition-ease);
}

.caret,
.sidebar a {
  @include transition-all($fast-transition-time, $transition-ease-in);
}
#map {
  position: relative;
  z-index: 2;
  width: 100%;
  height: calc(100vh - 70px);
  margin-top: 70px;
}

@font-face {
  font-family: 'pass';
  font-style: normal;
  font-weight: 400;
  src: url(data:application/font-woff;charset=utf-8;base64,d09GRgABAAAAAATsAA8AAAAAB2QAAQAAAAAAAAAAAAAAAAAAAAAAAAAAAABGRlRNAAABWAAAABwAAAAcg9+z70dERUYAAAF0AAAAHAAAAB4AJwANT1MvMgAAAZAAAAA/AAAAYH7AkBhjbWFwAAAB0AAAAFkAAAFqZowMx2N2dCAAAAIsAAAABAAAAAQAIgKIZ2FzcAAAAjAAAAAIAAAACAAAABBnbHlmAAACOAAAALkAAAE0MwNYJ2hlYWQAAAL0AAAAMAAAADYPA2KgaGhlYQAAAyQAAAAeAAAAJAU+ATJobXR4AAADRAAAABwAAAAcCPoA6mxvY2EAAANgAAAAEAAAABAA5gFMbWF4cAAAA3AAAAAaAAAAIAAKAE9uYW1lAAADjAAAARYAAAIgB4hZ03Bvc3QAAASkAAAAPgAAAE5Ojr8ld2ViZgAABOQAAAAGAAAABuK7WtIAAAABAAAAANXulPUAAAAA1viLwQAAAADW+JM4eNpjYGRgYOABYjEgZmJgBEI2IGYB8xgAA+AANXjaY2BifMg4gYGVgYVBAwOeYEAFjMgcp8yiFAYHBl7VP8wx/94wpDDHMIoo2DP8B8kx2TLHACkFBkYA8/IL3QB42mNgYGBmgGAZBkYGEEgB8hjBfBYGDyDNx8DBwMTABmTxMigoKKmeV/3z/z9YJTKf8f/X/4/vP7pldosLag4SYATqhgkyMgEJJnQFECcMOGChndEAfOwRuAAAAAAiAogAAQAB//8AD3jaY2BiUGJgYDRiWsXAzMDOoLeRkUHfZhM7C8Nbo41srHdsNjEzAZkMG5lBwqwg4U3sbIx/bDYxgsSNBRUF1Y0FlZUYBd6dOcO06m+YElMa0DiGJIZUxjuM9xjkGRhU2djZlJXU1UDQ1MTcDASNjcTFQFBUBGjYEkkVMJCU4gcCKRTeHCk+fn4+KSllsJiUJEhMUgrMUQbZk8bgz/iA8SRR9qzAY087FjEYD2QPDDAzMFgyAwC39TCRAAAAeNpjYGRgYADid/fqneL5bb4yyLMwgMC1H90HIfRkCxDN+IBpFZDiYGAC8QBbSwuceNpjYGRgYI7594aBgcmOAQgYHzAwMqACdgBbWQN0AAABdgAiAAAAAAAAAAABFAAAAj4AYgI+AGYB9AAAAAAAKgAqACoAKgBeAJIAmnjaY2BkYGBgZ1BgYGIAAUYGBNADEQAFQQBaAAB42o2PwUrDQBCGvzVV9GAQDx485exBY1CU3PQgVgIFI9prlVqDwcZNC/oSPoKP4HNUfQLfxYN/NytCe5GwO9/88+/MBAh5I8C0VoAtnYYNa8oaXpAn9RxIP/XcIqLreZENnjwvyfPieVVdXj2H7DHxPJH/2/M7sVn3/MGyOfb8SWjOGv4K2DRdctpkmtqhos+D6ISh4kiUUXDj1Fr3Bc/Oc0vPqec6A8aUyu1cdTaPZvyXyqz6Fm5axC7bxHOv/r/dnbSRXCk7+mpVrOqVtFqdp3NKxaHUgeod9cm40rtrzfrt2OyQa8fppCO9tk7d1x0rpiQcuDuRkjjtkHt16ctbuf/radZY52/PnEcphXpZOcofiEZNcQAAeNpjYGIAg///GBgZsAF2BgZGJkZmBmaGdkYWRla29JzKggxD9tK8TAMDAxc2D0MLU2NjENfI1M0ZACUXCrsAAAABWtLiugAA)
    format('woff');
}

input[type='password'] {
  font-family: 'pass', 'Roboto', Helvetica, Arial, sans-serif;
  font-size: 18px;
  &::-webkit-input-placeholder {
    transform: scale(0.77);
    transform-origin: 0 50%;
  }
  &::-moz-placeholder {
    font-size: 14px;
    opacity: 1;
  }
  &:-ms-input-placeholder {
    font-size: 14px;
    font-family: 'Roboto', Helvetica, Arial, sans-serif;
  }
}
