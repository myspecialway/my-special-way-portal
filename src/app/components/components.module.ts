import { NgModule } from '@angular/core';
import { CommonModule } from '@angular/common';
import { RouterModule } from '@angular/router';
import { FormsModule } from '@angular/forms';
import { MatProgressSpinnerModule } from '@angular/material/progress-spinner';
import { FooterComponent } from './footer/footer.component';
import { NavbarComponent } from './navbar/navbar.component';
import { ScheduleComponent } from './schedule/schedule.component';
import { SharedModule } from '../shared.module';
import { SpinnerComponent } from './spinner/spinner.component';
import { HttpClientModule } from '@angular/common/http';

@NgModule({
  imports: [CommonModule, RouterModule, FormsModule, MatProgressSpinnerModule, HttpClientModule],
  declarations: [FooterComponent, NavbarComponent, ScheduleComponent, SpinnerComponent],
  exports: [FooterComponent, NavbarComponent, ScheduleComponent, SpinnerComponent],
})
export class ComponentsModule {}
