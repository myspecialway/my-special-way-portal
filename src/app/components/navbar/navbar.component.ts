import { Component, OnDestroy, OnInit } from '@angular/core';
import { RouteInfo } from './models/route-info.model';
import { Apollo } from 'apollo-angular';
import { GET_USER_PROFILE } from '../../apollo/state/queries/get-user-profile.query';
import { UserProfileStateModel } from '../../apollo/state/resolvers/state.resolver';
import { UserType } from '../../models/user.model';
import { Router, NavigationEnd } from '@angular/router';
import { pluck, filter, map } from 'rxjs/operators';
import { Subscription } from 'rxjs';
import { TranslateService } from '@ngx-translate/core';

export const ROUTES: RouteInfo[] = [
  { path: 'student', title: 'ROUTES.STUDENT', class: 'nb-student', roles: [UserType.PRINCIPLE, UserType.TEACHER] },
  { path: 'class', title: 'ROUTES.CLASSES', class: 'nb-class', roles: [UserType.PRINCIPLE] },
  { path: 'lesson', title: 'ROUTES.LESSON', class: 'nb-lesson', roles: [UserType.PRINCIPLE] },
  { path: 'class/:id', title: 'ROUTES.CLASS', class: 'nb-class-schedule', roles: [UserType.TEACHER] },
  { path: 'map', title: 'ROUTES.MAP', class: 'nb-map', roles: [UserType.PRINCIPLE] },
  { path: 'user', title: 'ROUTES.USER', class: 'nb-user', roles: [UserType.PRINCIPLE] },
];

export const DEFAULT_ROUTE = ROUTES[0];

@Component({
  selector: 'app-navbar',
  templateUrl: './navbar.component.html',
  styleUrls: ['./navbar.component.scss'],
})
export class NavbarComponent implements OnInit, OnDestroy {
  menuItems: any[];
  currentUser: string;
  selectedMenuItemPath: string;
  routeSubscription: Subscription;

  constructor(private apollo: Apollo, private router: Router, public translate: TranslateService) {
    this.subscribeToRouterEvents();

    ROUTES.forEach((route) => {
      this.translate.get(route.title).subscribe((resource) => {
        route.title = resource;
      });
    });
  }

  async ngOnInit() {
    this.apollo
      .watchQuery<{ userProfile: UserProfileStateModel }>({
        query: GET_USER_PROFILE,
      })
      .valueChanges.subscribe((userProf) => {
        this.currentUser = `${userProf.data.userProfile.firstname} ${userProf.data.userProfile.lastname}`;
        const currentType = userProf.data.userProfile.role;
        const currentClassId = userProf.data.userProfile.class_id;
        this.menuItems = ROUTES.filter((menuItem) => menuItem.roles.includes(UserType[currentType]));
        this.menuItems.forEach((item) => {
          if (item.path.indexOf('/:id') >= 0) {
            item.path = item.path.replace('/:id', currentClassId ? '/' + currentClassId : '');
          }
          return item;
        });
      });
  }

  ngOnDestroy() {
    this.routeSubscription.unsubscribe();
  }

  selectMenuItem(menuItem) {
    this.selectedMenuItemPath = menuItem.path;
  }

  getSelectedMenuItem() {
    const route = ROUTES.find((menuItem) => menuItem.path === this.selectedMenuItemPath) || DEFAULT_ROUTE;
    return route.title;
  }

  private subscribeToRouterEvents() {
    this.routeSubscription = this.router.events
      .pipe(
        filter((ev) => ev instanceof NavigationEnd),
        pluck('url'),
        map(this.removeLeadingSlash),
      )
      .subscribe((path) => (this.selectedMenuItemPath = path));
  }

  private removeLeadingSlash(s = '') {
    return s.replace(/^\/+/g, '');
  }
}
