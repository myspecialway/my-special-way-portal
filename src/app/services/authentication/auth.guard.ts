import gql from 'graphql-tag';
import { Injectable } from '@angular/core';
import { ActivatedRouteSnapshot, CanActivate, Router, RouterStateSnapshot, UrlSegment } from '@angular/router';
import { AuthenticationService } from './authentication.service';
import { UserType } from '../../models/user.model';
import { Apollo } from 'apollo-angular';
import { UserProfileStateModel } from '../../apollo/state/resolvers/state.resolver';

@Injectable()
export class AuthGuard implements CanActivate {
  constructor(
    private router: Router,
    private authService: AuthenticationService,
    private apollo: Apollo,
  ) { }

  async canActivate(route: ActivatedRouteSnapshot, state: RouterStateSnapshot): Promise<boolean> {
    await this.authService.checkRestoreAuthData();
    const response = await this.apollo.query<{ userProfile: UserProfileStateModel }>({
      query: gql`
      query {
        userProfile @client{
          role
          token
        }
      }
    `}).toPromise();

    if (
      response.data.userProfile &&
      response.data.userProfile.token &&
      !this.authService.isTokenExpired(response.data.userProfile.token)
      && this.isAuthorized(route.url, response.data.userProfile.role)
    ) {
      return true;
    }

    this.router.navigate(['/login'], { queryParams: { returnUrl: state.url } });
    return false;
  }

  isAuthorized(route: UrlSegment[], userRole: UserType) {
    // console.log(`route ${route} | User type ${userRole}`);
    // if (route[0].path.toString() === 'user' && userRole.toString() === 'TEACHER') {
    //   return false;
    // }
    return true; // TODO: authorization logic
  }
}
