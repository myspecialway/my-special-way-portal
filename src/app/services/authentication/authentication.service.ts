import 'rxjs/add/operator/map';
import 'rxjs/add/operator/catch';
import { Injectable } from '@angular/core';
import { HttpClient, HttpErrorResponse } from '@angular/common/http';
import { LoginResponse } from '../../models/login-response.model';
import { environment } from '../../../environments/environment';
import { JwtHelperService } from '@auth0/angular-jwt';
import { JWTTokenPayloadResponse } from '../../models/jwt-token-resonse.model';
import { Apollo } from 'apollo-angular';
import { UPDATE_USER_PROFILE } from '../../apollo/state/mutations/update-user-profile.mutation';
import { UserProfileStateModel } from '../../apollo/state/resolvers/state.resolver';
import { UserUniqueValidationResponse } from '../../models/user-unique-validation-response.model';
import { of } from 'rxjs';
import { catchError, first, map } from 'rxjs/operators';

@Injectable()
export class AuthenticationService {
  private jwtHelper = new JwtHelperService();
  private isStateRestored = false;

  constructor(private http: HttpClient, private apollo: Apollo) {}

  private getTokenFromLocalStore() {
    return localStorage.getItem('token') ? localStorage.getItem('token') : sessionStorage.getItem('token');
  }

  async checkRestoreAuthData() {
    if (this.isStateRestored) {
      return;
    }

    this.isStateRestored = true;
    const token = this.getTokenFromLocalStore();

    if (!token) {
      return;
    }

    const userProfile = this.getProfileFromToken(token);
    await this.pushUserProfileToState(userProfile);
  }

  async login(username: string, password: string, isRememberMeActive: boolean): Promise<boolean> {
    try {
      const tokenResponse = await this.http
        .post<LoginResponse>(environment.hotConfig.MSW_HOT_LOGIN_ENDPOINT, { username, password })
        .toPromise();

      if (!tokenResponse) {
        return false;
      }

      if (isRememberMeActive) {
        localStorage.setItem('token', tokenResponse.accessToken);
      } else {
        sessionStorage.setItem('token', tokenResponse.accessToken);
      }

      const userProfile = this.getProfileFromToken(tokenResponse.accessToken);
      await this.pushUserProfileToState(userProfile);

      return true;
    } catch (error) {
      const typedError = error as HttpErrorResponse;

      if (typedError.status !== 401) {
        throw error;
      }

      return false;
    }
  }

  private getProfileFromToken(token: string): UserProfileStateModel {
    const jwrParsedToken = this.parseToken(token);
    const userProfile = new UserProfileStateModel(jwrParsedToken);
    userProfile.token = token;

    return userProfile;
  }

  private parseToken(token: string): JWTTokenPayloadResponse {
    const decodedToken = this.jwtHelper.decodeToken(token) as JWTTokenPayloadResponse;
    return decodedToken;
  }

  private async pushUserProfileToState(userProfile: UserProfileStateModel) {
    return this.apollo
      .mutate({
        mutation: UPDATE_USER_PROFILE,
        variables: {
          userProfile,
        },
      })
      .toPromise();
  }

  isTokenExpired(token: string): boolean {
    return this.jwtHelper.isTokenExpired(token);
  }

  logout() {
    localStorage.removeItem('token');
    sessionStorage.removeItem('token');
  }

  checkUsernameUnique(username: string, id: string) {
    if (!username) {
      return of(true);
    }
    const userID = id === '' ? undefined : id;
    try {
      return this.http
        .post<UserUniqueValidationResponse>(environment.hotConfig.MSW_HOT_USER_UNIQUE_VALIDATION_ENDPOINT, {
          username,
          id: userID,
        })
        .pipe(
          first(),
          map((res) => res.isUnique),
          catchError(() => of(true)),
        );
    } catch (error) {
      return of(true);
    }
  }
}
