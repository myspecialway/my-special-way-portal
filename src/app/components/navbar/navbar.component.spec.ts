import { NavbarComponent } from './navbar.component';
import { TestBed, tick, fakeAsync } from '@angular/core/testing';
import { NO_ERRORS_SCHEMA, Injectable } from '@angular/core';
import { AuthenticationService } from '../../services/authentication/authentication.service';
import { HttpClient, HttpHandler } from '@angular/common/http';
import { Shallow } from 'shallow-render/dist';
import { ComponentsModule } from '../components.module';
import { Apollo } from 'apollo-angular';
import { Subject } from 'rxjs';
import { Router, NavigationEnd } from '@angular/router';
import { TranslateModule, TranslateLoader } from '@ngx-translate/core';
import { TranslateCustomLoader } from '../../../mocks/translate.stub';

const fixUrl = (s) => s.replace(/^\/+/g, '');

@Injectable()
export class RouterStub {
  url: string;

  private subject = new Subject();
  events = this.subject.asObservable();

  navigate([url]: string[]) {
    this.url = url;
    this.triggerNavEvents(url);
  }

  triggerNavEvents(url) {
    const ne = new NavigationEnd(0, url, url);
    this.subject.next(ne);
  }
}

describe('navbar component', () => {
  let shallow: Shallow<NavbarComponent>;
  let watchQueryObservable: Subject<any>;
  let router: RouterStub;

  beforeEach(async () => {
    watchQueryObservable = new Subject();
    const apolloMock = {
      watchQuery: () => ({
        valueChanges: watchQueryObservable,
      }),
    };
    shallow = new Shallow(NavbarComponent, ComponentsModule);
    TestBed.configureTestingModule({
      declarations: [NavbarComponent],
      providers: [
        { provide: Apollo, useValue: apolloMock },
        HttpClient,
        HttpHandler,
        AuthenticationService,
        { provide: Router, useClass: RouterStub },
      ],
      imports: [
        TranslateModule.forRoot({
          loader: {
            provide: TranslateLoader,
            useClass: TranslateCustomLoader,
          },
        }),
      ],
      schemas: [NO_ERRORS_SCHEMA],
    });
    // const apollo = TestBed.get(HttpClient);
  });
  it('should render component as described in snapshot', () => {
    const fixture = TestBed.createComponent(NavbarComponent);
    expect(fixture).toMatchSnapshot();
  });

  it('should keep selected menu item path reflecting the route path', fakeAsync(() => {
    // given
    router = TestBed.get(Router);
    const fixture = TestBed.createComponent(NavbarComponent);
    expect(fixture.componentInstance.selectedMenuItemPath).toBeUndefined();

    // router.initialNavigation();
    const givenRoutePath = '/test-path';
    const expectedMenuItemPath = fixUrl(givenRoutePath);
    // when
    router.navigate([givenRoutePath]);
    tick();

    // then
    expect(fixture.componentInstance.selectedMenuItemPath).toBe(expectedMenuItemPath);
  }));

  it('should render username per authentication service on init - for principle', async () => {
    // given
    const component = await shallow.render('<app-navbar></app-navbar>');

    // when
    watchQueryObservable.next({
      data: {
        userProfile: {
          username: 'test',
          firstname: 'MSW',
          lastname: 'PRINCIPLE',
          role: 'PRINCIPLE',
          class_id: 'some_classid',
        },
      },
    });

    // then
    component.fixture.detectChanges();
    const liElement = component.element.nativeElement.querySelector('.msw-header-user-name') as HTMLLIElement;
    expect(liElement.innerHTML).toBe('MSW PRINCIPLE');
  });

  it('should update the selected menu item on selection', async () => {
    const fixture = TestBed.createComponent(NavbarComponent);
    fixture.componentInstance.selectMenuItem({ path: 'class', title: 'ROUTES.CLASSES', class: 'nb-class' });
    expect(fixture.componentInstance.selectedMenuItemPath).toEqual('class');
  });

  it('should return the correct menu item', () => {
    const fixture = TestBed.createComponent(NavbarComponent);
    fixture.componentInstance.selectMenuItem({ path: 'class', title: 'ROUTES.CLASSES', class: 'nb-class' });
    const title = fixture.componentInstance.getSelectedMenuItem();
    expect(title).toEqual('ROUTES.CLASSES');
  });

  it('should render username per authentication service on init - for teacher', async () => {
    // given
    const component = await shallow.render('<app-navbar></app-navbar>');

    // when
    watchQueryObservable.next({
      data: {
        userProfile: {
          username: 'test',
          firstname: 'MSW',
          lastname: 'TEACHER',
          role: 'TEACHER',
          class_id: 'some_classid',
        },
      },
    });

    // then
    component.fixture.detectChanges();
    const liElement = component.element.nativeElement.querySelector('.msw-header-user-name') as HTMLLIElement;
    expect(liElement.innerHTML).toBe('MSW TEACHER');
  });
});
