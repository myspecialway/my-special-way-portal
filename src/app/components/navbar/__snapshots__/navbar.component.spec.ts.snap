// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`navbar component should render component as described in snapshot 1`] = `
<app-navbar
  apollo={[Function Object]}
  routeSubscription={[Function Subscriber]}
  router={[Function RouterStub]}
  translate={[Function TranslateService]}
>
  <nav
    class="navbar navbar-expand-lg navbar-white  navbar-absolute fixed-top"
  >
    <div
      class="container-fluid"
    >
       
      <div
        class="navbar-wrapper"
      >
        <a
          class="navbar-brand"
          href="https://www.tlv-edu.gov.il/sites/yachdav/Pages/Vision.aspx"
        >
          <img
            class="logo-img"
            src="/assets/img/msw-logo.png"
          />
          <img
            class="logo-img-tlv"
            src="/assets/img/tlv-logo-small-white.jpg"
          />
        </a>
        <div
          class="msw-page-title-dropdown dropdown"
        >
          <a
            aria-expanded="false"
            aria-haspopup="true"
            class=""
            data-toggle="dropdown"
            href="http://example.com"
            id="navbarDropdownMenuLink0"
          >
            <span
              class="msw-page-title"
            >
              <span>
                
              </span>
              <i
                class="material-icons"
              >
                arrow_drop_down
              </i>
            </span>
          </a>
          <div
            aria-labelledby="navbarDropdownMenuLink"
            class="dropdown-menu dropdown-menu-link"
          >
            
          </div>
        </div>
      </div>
      <div
        class="collapse navbar-collapse justify-content-end"
        id="navigation"
      >
        
      </div>
    </div>
  </nav>
</app-navbar>
`;
